### ISSUE TITLE
Support: Git Repos and Automation  

### ISSUE THREAD

**Issue by juliaford** [thinkshout/bene#280]
## Overview

Bene client repositories are not structured like our other, modern Drupal 8 sites which makes them difficult to work on. Whatever solution we land on, we’d like to have the source repository for Bene clients only include custom modules and themes.

There are two ideas for solutions:

1. Continue to use Pantheon upstreams and use a combination of Composer and a build script to setup your local site and push to Pantheon. The positive here is that you could still merge in the upstream, but the negative is that this is non-standard and is not compatible with per-site Composer dependencies or patches.
1. Move completely to Composer and drop Pantheon upstreams. The positives here are that Bene sites would match the industry standard for Drupal 8, and different sites could use different (patch, non-security) versions of Bene which is difficult if not impossible with upstreams. The negative is that you couldn’t “one click update” in Pantheon.

I lean heavily towards (2) - updating and working on Bene sites is already difficult and we use upstreams today, so I think moving us to a modern build makes sense. This also matches how other distributions do their builds - most people just use Composer and do not have overly complex install scripts for building the codebase. Updating a Bene site using Composer would mean checking out the repo, running “composer update drupal/bene”, pushing that change, and letting CircleCI build and deploy to Pantheon.

We'll need to decide what path to take before starting on the issue.

## This ticket is done when:

TBD.

**Comment by mortenson** [comment]
We're going to move forward with moving to Composer by default, but continuing to maintain the upstream. For update automation, we're figuring something like the following:
1. A new Bene release comes out
2. CircleCI responds to the new release (or is manually triggered)
3. CI checks out the Github repo, creates a branch from master, and runs `composer update drupal/bene --with-dependencies`
4. Commit the changes and push to a Github repo branch (could be a consistent name like `auto-update` or generated like `update-VERSION`)
5. Another CI build runs in response to the new branch, creates/deploys to a multidev
6. That CI build also runs a visual regression test, standard tests
7. Testing results are (hopefully quickly) reviewed and the new branch is merged into master
8. Update goes to live

**Comment by unclegcb** [comment]
Step 2 here, by the way -- that's CircleCI on each individual Bene site responding to the release on the Bene repo, right? Not a single circle configuration responding by building multiple sites?

**Comment by mortenson** [comment]
@unclegcb That's correct - we would do a CircleCI run per site.

**Comment by mortenson** [comment]
Note to document:
1. Setting up a new site from scratch with Pantheon
2. Going back to the upstream from bene-project

**Comment by mortenson** [comment]
Here's the process I've been using to move a site from the upstream onto the bene-project structure:
1. Run `git clone git@github.com:thinkshout/bene-project.git ~/repos/bene-project` to get a local copy of the bene-project.
2. Run `cd ~/repos/bene-project && composer install` to scaffold out what a new site may look like.
3. Clone the current site repo somewhere and `cd` into it.
4. Run `git checkout reporedux`
5. Run `comm -23 <(find web/modules/contrib -type d -depth 1 | sort) <(find ~/repos/bene-project/web/modules/contrib -type d -depth 1 | sed 's/.*web/web/' | sort)` to get a list of modules that are in the site repo that differ from bene-project. Write down list including module verisons.
6. Run the following:
```
git rm -rf *
rm .travis.yml .gitignore .env.dist
cp -r ~/repos/bene-project/* .
cp ~/repos/bene-project/.gitignore .
cp ~/repos/bene-project/.env.dist .
git checkout master -- web/themes
git checkout master -- web/sites/default
```
7. Run `composer require` for all the modules that need re-added based on what you recorded earlier.
8. Run `cp -r ~/repos/bene-project/.circleci .circleci` and follow the instructions in `~/repos/bene-project/.circleci` to configure CircleCI.
9. Comment out the "deploy" section of the circleci config file.
10. Run `composer install` to create a lock file if one does not exist.
11. Audit deletions to find files that should not have been removed. One command to make this easier is `git status | grep deleted: | grep -v profiles | grep -v modules | grep -v core | grep -v vendor`. Checkout files that should be present.
12. Check if the theme imports files from a bene profile theme, and update the import paths if necessary then run `rake serve` to compile new CSS.
13. Create a `reporedux` multidev in Pantheon.
14. Commit/push and make sure the CircleCI build finishes without errors.
15. Run the following to fix a bug related to moving the profile, where NAME is the Pantheon site name:
```
drush @pantheon.NAME.reporedux php-script scripts/fixpath.php
drush @pantheon.NAME.reporedux cr
```
16. Make a PR from `reporedux` to `master` and send for QA.

**Comment by unclegcb** [comment]
Tried instructions on OTFO and got stuck on step 4: getting a strange error in Circle: https://circleci.com/gh/thinkshout/otfo-bene/4

Did I manage to hustle up some broken package?

**Comment by mortenson** [comment]
@unclegcb It looks like `vendor`, `web/core`, and `web/modules/contrib` still exist for that repo. Maybe you didn't commit all the removals from `rm -rf *`?

**Comment by unclegcb** [comment]
Good point! I think this is because they get stuck in git. Our instructions run `composer install` which makes these files present in the `bene-project` directory. So, after we delete them in our reporedux branch, we just copy new ones in when we `cp -r ~/repos/bene-project/* `.

Since our primary repo had those files in .git, this flummoxes .gitignore. We need to manually `git rm` them, I think, before running `cp -r ~/repos/bene-project/* `. Alternately, we need to copy bene-project from a version of bene-project that has never had `composer install` run.

**Comment by unclegcb** [comment]
It looks to me like we need to change `rm -rf *` to `git rm -r *`. Testing now.

**Comment by unclegcb** [comment]
Success! https://circleci.com/gh/thinkshout/otfo-bene/7

**Comment by unclegcb** [comment]
Ok, another step to add: update the include paths in the child theme that point to bene_parent.

**Comment by unclegcb** [comment]
This is in good shape. Update by site:

Freegeek: nothing has been done, no intent to change their process at this point, but we might want to discuss with them.
CORI: https://test-cori.pantheonsite.io/ has the updated system. It is ready to deploy to live with approval. https://github.com/thinkshout/cori-bene/issues/6
WRI: https://test-wri-bene.pantheonsite.io/ has the updated system. It is ready to deploy to live with approval.https://github.com/thinkshout/wri-bene/issues/4
STTR: Minor issue was obvious on the test site, ticket [here](https://github.com/thinkshout/streettrust-bene/pull/119#issuecomment-477816260) 
OTFO: Updated
WEAP: New system from the get-go
ULPD: New system from the get-go

**Comment by mshaver** [comment]
I'm assigning QA for the CORI and WRI sites to me for final approval. Note @unclegcb, we need to update salsa templates for CORI whenever we do a release that rebuilds the sites CSS files. Unfortunate for sure, but not something we're going to solve here.

**Comment by mshaver** [comment]
@unclegcb, both CORI and WIRI are good to go from my QA. I'll reach out to clients and set a date on the calendar for next week to deploy these.

**Comment by mshaver** [comment]
Closing
